### PR TITLE
EIP-1559: miner changes

### PR DESCRIPTION
**Pull request by holiman**
This PR contains the changes for the miner, taken from https://github.com/ethereum/go-ethereum/pull/22833 , and lightly squashed. 
This PR is based on `master`, and does not require the txpool changes to be merged first. 

Supersedes  https://github.com/ethereum/go-ethereum/pull/22833

### CODE REVIEW

File: core/types/transaction.go [hunk 1]
```diff
@@ -299,6 +300,19 @@ func (tx *Transaction) Cost() *big.Int {
 	return total
 }
 
+// EffectiveTip returns the effective miner tip for the given base fee.
+// Returns error in case of a negative effective miner tip.
+func (tx *Transaction) EffectiveTip(baseFee *big.Int) (*big.Int, error) {
```
Comment (zsfelfoldi) [inline core/types/transaction.go:305]: This is correct but I also added this function with slightly different semantics in my tx pool PR: my version does not return an error, it returns the negative value which the caller can handle as it wishes. For me this value is useful even if negative (I described the reasons in my gist). Do you think we should have two versions? Or use my version and check at the caller side when necessary? I guess you should check it for the minimum effective fee threshold anyway.

**Review by zsfelfoldi — APPROVED** [review]
LGTM

File: core/types/transaction.go [hunk 2]
```diff
@@ -400,24 +414,44 @@ func (s TxByNonce) Len() int           { return len(s) }
...
+}
+
 // TxByPriceAndTime implements both the sort and the heap interface, making it useful
 // for all at once sorting as well as individually adding and removing elements.
-type TxByPriceAndTime Transactions
+type TxByPriceAndTime []*TxWithMinerFee
 
 func (s TxByPriceAndTime) Len() int { return len(s) }
 func (s TxByPriceAndTime) Less(i, j int) bool {
 	// If the prices are equal, use the time the transaction was first seen for
```
Comment (MariusVanDerWijden) [inline core/types/transaction.go:443]: Comment is outdated

**Review by MariusVanDerWijden — APPROVED** [review]
LGTM, though I could not meaningfully review the `CalcGasLimit1559` function

File: core/types/transaction.go (hunk 1 above)
Comment (holiman) [reply core/types/transaction.go:305]: I think maybe let's start with this. We could have this method maybe wrap your method, so the 'normal' case doesn't need to handle negative tips

**Comment by Mohdhasni99** [comment]
Ok

**Pull request merged**